source/base/qprojector.cc

```diff
@@ -1140,48 +1140,107 @@ Quadrature<2>
 
-  Assert(reference_cell == ReferenceCells::Quadrilateral, ExcNotImplemented());
+  const unsigned int n_points          = quadrature.size(),
+                     n_faces           = reference_cell.n_faces(),
+                     subfaces_per_face = reference_cell.face_reference_cell(0)
+                                           .n_isotropic_children();
 
-  const unsigned int dim = 2;
+  if (reference_cell == ReferenceCells::Quadrilateral)
+    {
+      // first fix quadrature points
+      std::vector<Point<dim>> q_points;
+      q_points.reserve(n_points * n_faces * subfaces_per_face);
+      std::vector<Point<dim>> help(n_points);
 
-  const unsigned int n_points = quadrature.size(),
-                     n_faces  = GeometryInfo<dim>::faces_per_cell,
-                     subfaces_per_face =
-                       GeometryInfo<dim>::max_children_per_face;
+      // project to each face and copy
+      // results
+      for (unsigned int face = 0; face < n_faces; ++face)
+        for (unsigned int subface = 0; subface < subfaces_per_face; ++subface)
+          {
+            project_to_subface(reference_cell, quadrature, face, subface, help);
+            std::copy(help.begin(), help.end(), std::back_inserter(q_points));
+          }
 
-  // first fix quadrature points
-  std::vector<Point<dim>> q_points;
-  q_points.reserve(n_points * n_faces * subfaces_per_face);
-  std::vector<Point<dim>> help(n_points);
+      // next copy over weights
+      std::vector<double> weights;
+      weights.reserve(n_points * n_faces * subfaces_per_face);
+      for (unsigned int face = 0; face < n_faces; ++face)
+        for (unsigned int subface = 0; subface < subfaces_per_face; ++subface)
+          std::copy(quadrature.get_weights().begin(),
+                    quadrature.get_weights().end(),
+                    std::back_inserter(weights));
 
-  // project to each face and copy
-  // results
-  for (unsigned int face = 0; face < n_faces; ++face)
-    for (unsigned int subface = 0; subface < subfaces_per_face; ++subface)
-      {
-        project_to_subface(reference_cell, quadrature, face, subface, help);
-        std::copy(help.begin(), help.end(), std::back_inserter(q_points));
-      }
+      Assert(q_points.size() == n_points * n_faces * subfaces_per_face,
+             ExcInternalError());
+      Assert(weights.size() == n_points * n_faces * subfaces_per_face,
+             ExcInternalError());
 
-  // next copy over weights
-  std::vector<double> weights;
-  weights.reserve(n_points * n_faces * subfaces_per_face);
-  for (unsigned int face = 0; face < n_faces; ++face)
-    for (unsigned int subface = 0; subface < subfaces_per_face; ++subface)
-      std::copy(quadrature.get_weights().begin(),
-                quadrature.get_weights().end(),
-                std::back_inserter(weights));
+      return Quadrature<dim>(std::move(q_points), std::move(weights));
+    }
+  else if (reference_cell == ReferenceCells::Triangle)
```

Comment:
Wouldn't it be possible to unify this code with https://github.com/dealii/dealii/blob/26d80990ce90c702173339c7f762047c79286f2e/source/base/qprojector.cc#L773. There is a lot of duplication here.